src/inbox/chan_ptr.rs

```diff
@@ -0,0 +1,402 @@
+/// out more.
+#[derive(Debug)]
+pub struct TxStrong(());
+
+impl TxStrong {
+    /// Attempt to construct a new `TxStrong` pointing to the given `inner` if there are existing
+    /// strong references to `inner`. This will return `None` if there were 0 strong references to
+    /// the inner.
+    fn try_new<A>(inner: &Chan<A>) -> Option<TxStrong> {
+        // All code taken from Weak::upgrade in std
+        use std::sync::atomic::Ordering::*;
+
+        // Relaxed load because any write of 0 that we can observe leaves the field in a permanently
+        // zero state (so a "stale" read of 0 is fine), and any other value is confirmed via the
+        // CAS below.
+        let mut n = inner.sender_count.load(Relaxed);
+
+        loop {
+            if n == 0 {
+                return None;
+            }
+
+            // Relaxed is fine for the failure case because we don't have any expectations about the new state.
+            // Acquire is necessary for the success case to synchronise with `Arc::new_cyclic`, when the inner
+            // value can be initialized after `Weak` references have already been created. In that case, we
+            // expect to observe the fully initialized value.
+            match inner
+                .sender_count
+                .compare_exchange_weak(n, n + 1, Acquire, Relaxed)
+            {
+                Ok(_) => return Some(TxStrong(())), // 0-case checked above
+                Err(old) => n = old,
+            }
+        }
+    }
+}
+
+/// The reference count of a weak address. Weak addresses will bit prevent the actor from being
+/// dropped. Read the docs of [`Address`](crate::Address) to find out more.
+#[derive(Debug)]
+pub struct TxWeak(());
+
+/// A reference counter that can be dynamically either strong or weak.
+#[derive(Debug)]
+pub enum TxEither {
+    /// A strong reference counter.
+    Strong(TxStrong),
+    /// A weak reference counter.
+    Weak(TxWeak),
+}
+
+/// A reference counter for the receiving end of the channel or in actor terminology, the mailbox
+/// of an actor.
+pub struct Rx(());
+
+/// Defines a reference counting policy for a channel pointer.
+///
+/// # Implementation note
+///
+/// This trait only exists because we cannot specialise `Drop` impls in Rust, otherwise, we could
+/// write:
+///
+/// - `impl Drop for ChanPtr<A, TxStrong> { }`
+/// - `impl Drop for ChanPtr<A, TxEither> { }`
+/// - `impl Drop for ChanPtr<A, Rx> { }`
+///
+/// and call the appropriate functions on [`Chan`].
+pub trait RefCounter: Send + Sync + 'static + Unpin + private::RefCounter {}
+
+impl RefCounter for TxStrong {}
+impl RefCounter for TxWeak {}
+impl RefCounter for TxEither {}
+impl RefCounter for Rx {}
+
+/// Private module to ensure [`RefCounter`] is sealed and can neither be implemented on other types
+/// nor can a user outside of this crate call functions from [`RefCounter`].
+mod private {
+    use super::*;
+
+    pub trait RefCounter {
+        type Either;
```

Comment:
Can this not directly refer to the `Either` type instead of being through an assoc type?

Reply:
It can although it isn't super clean because `RefCounter` is more generic than `TxEither`. We can name the function `into_tx_either` and panic similar as we do now to avoid the associated type.